dipy/reconst/mapmri.py

```diff
@@ -135,74 +256,154 @@ def __init__(self,
+            elif type(self.laplacian_weighting) == np.ndarray:
+                lopt = generalized_crossvalidation_array(data, M,
+                                    laplacian_matrix, self.laplacian_weighting)
+        else:
+            lopt = 0.
+            laplacian_matrix = np.ones((self.ind_mat.shape[0],
+                                        self.ind_mat.shape[0]))
 
-        qvals = np.sqrt(self.gtab.bvals / self.tau) / (2 * np.pi)
-        qvecs = np.dot(self.gtab.bvecs, R)
-        q = qvecs * qvals[:, None]
-        M = mapmri_phi_matrix(self.radial_order, mu, q.T)
-        # This is a simple empirical regularization, to be replaced
-        I = np.diag(self.ind_mat.sum(1) ** 2)
-        if self.eap_cons:
-            if not have_cvxopt:
-                raise ValueError(
-                    'CVXOPT package needed to enforce constraints')
-            w_s = "The implementation of MAPMRI depends on CVXOPT "
-            w_s += " (http://cvxopt.org/). This software is licensed "
+        if self.positivity_constraint:
+            w_s = "The MAPMRI positivity constraint depends on CVXOPT "
+            w_s += "(http://cvxopt.org/). CVXOPT is licensed "
             w_s += "under the GPL (see: http://cvxopt.org/copyright.html) "
-            w_s += " and you may be subject to this license when using MAPMRI."
+            w_s += "and you may be subject to this license when using the "
+            w_s += "positivity constraint."
             warn(w_s)
-            import cvxopt.solvers
-            rmax = 2 * np.sqrt(10 * evals.max() * self.tau)
-            r_index, r_grad = create_rspace(11, rmax)
-            K = mapmri_psi_matrix(
-                self.radial_order,  mu, r_grad[0:len(r_grad) / 2, :])
-
-            Q = cvxopt.matrix(np.dot(M.T, M) + self.lambd * I)
-            p = cvxopt.matrix(-1 * np.dot(M.T, data))
+            if self.pos_radius == 'adaptive':
+                # custom constraint grid based on scale factor [Avram2015]
+                constraint_grid = create_rspace(self.pos_grid,
+                                                np.sqrt(5) * mu_max)
+            else:
+                constraint_grid = self.constraint_grid
+            if self.anisotropic_scaling:
+                K = mapmri_psi_matrix(self.radial_order, mu, constraint_grid)
+            else:
+                if self.pos_radius == 'adaptive':
+                    # grid changes per voxel. Recompute entire K matrix.
+                    K = mapmri_isotropic_psi_matrix(self.radial_order, mu[0],
+                                                    constraint_grid)
+                else:
+                    # grid is static. Only compute mu-dependent part of K.
+                    K_dependent = mapmri_isotropic_K_mu_dependent(
+                        self.radial_order, mu[0], constraint_grid)
+                    K = K_dependent * self.pos_K_independent
+
+            data = np.asarray(data / data[self.gtab.b0s_mask].mean())
```

Comment:
Should we add a `np.isscalar` on the output here @matthew-brett ? It would ber really ugly to special-case this to what is essentially just a bug in one version of Numpy.